Library/Homebrew/rubocops/components_order.rb

```diff
@@ -49,50 +51,123 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
           ]
 
-          @present_components = component_precedence_list.map do |components|
-            relevant_components = []
-            components.each do |component|
-              case component[:type]
-              when :method_call
-                relevant_components += find_method_calls_by_name(body_node, component[:name]).to_a
-              when :block_call
-                relevant_components += find_blocks(body_node, component[:name]).to_a
-              when :method_definition
-                relevant_components << find_method_def(body_node, component[:name])
+          @present_components, @offensive_nodes = check_order(component_precedence_list, body_node)
+
+          component_problem @offensive_nodes[0], @offensive_nodes[1] if @offensive_nodes
+
+          component_precedence_list = [
+            [{ name: :depends_on, type: :method_call }],
+            [{ name: :resource, type: :block_call }],
+            [{ name: :patch, type: :method_call }, { name: :patch, type: :block_call }],
+          ]
+
+          on_macos_blocks = find_blocks(body_node, :on_macos)
+
+          problem ": there can only be one on_macos block in a formula." if on_macos_blocks.length > 1
+
+          unless on_macos_blocks.empty?
+            on_macos_block = on_macos_blocks[0]
+            _, offensive_node = check_order(component_precedence_list, on_macos_block.body)
+            component_problem offensive_node[0], offensive_node[1] if offensive_node
+            on_macos_block.body.child_nodes.each do |child|
+              valid_node = false
+              valid_node = true if depends_on_node?(child)
+              next unless child.instance_of? RuboCop::AST::SendNode
+
+              valid_node = true if child.method_name.to_s == "patch"
+              valid_node = true if child.method_name.to_s == "resource"
+
+              @offensive_node = on_macos_block
+              @offense_source_range = on_macos_block.source_range
+              unless valid_node
+                problem "`on_macos` block can only include `depends_on`, `patch` and `resource` nodes."
               end
             end
-            relevant_components.delete_if(&:nil?)
           end
 
-          # Check if each present_components is above rest of the present_components
-          @present_components.take(@present_components.size - 1).each_with_index do |preceding_component, p_idx|
-            next if preceding_component.empty?
+          on_linux_blocks = find_blocks(body_node, :on_linux)
 
-            @present_components.drop(p_idx + 1).each do |succeeding_component|
-              next if succeeding_component.empty?
+          problem ": there can only be one on_linux block in a formula." if on_linux_blocks.length > 1
 
-              @offensive_nodes = check_precedence(preceding_component, succeeding_component)
-              component_problem @offensive_nodes[0], @offensive_nodes[1] if @offensive_nodes
+          unless on_linux_blocks.empty?
```

Comment:
```suggestion
          unless on_linux_blocks.any?
```

Reply:
This looks like an almost exact duplicate of the `on_macos_blocks`. Any chance of reducing the duplication?